File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
let visited = [];
let tree = [];
let N = 0;
let answer = Number.MAX_SAFE_INTEGER;

function dfs(now){
    visited[now] = true;
    
    let sum = 0;
    for(let x of tree[now]){
        if(!visited[x]){
            sum += dfs(x);
        }
    }
    
    sum++;
    answer = Math.min(answer, Math.abs(sum - (N-sum)));
    return sum;
}

function solution(n, wires) {
    N = n;
    visited = new Array(n).fill(false);
    tree = Array.from({length: n}, () => []);
    
    for(let [a, b] of wires){
        tree[a-1].push(b-1);
        tree[b-1].push(a-1);
    }
    
    dfs(0);
    
    return answer;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    graph = [[] for _ in range(n)]
    visited = [False] * n
    answer = int(1e9)
    
    for x in wires:
        graph[x[0]-1].append(x[1]-1)
        graph[x[1]-1].append(x[0]-1)
        
    def dfs(now):
        nonlocal answer
        
        sum = 1
        for x in graph[now]:
            if not visited[x]:
                visited[x] = True
                sum += dfs(x)
           

        answer = min(abs(sum - (n-sum)), answer)
        return sum

        
    for i in range(n):
        if not visited[i]:
            dfs(i)
            
    return answer
    
    